### PR TITLE
[Core] Add default-off preemption_victim="lcf" (least-computed-first) victim policy

### PR DESCRIPTION
**Pull request by shivasathishs-rp**
## What

Adds a new, independent scheduler axis `preemption_victim` that controls
*which* running request is evicted when the scheduler must preempt to free
KV cache. It is orthogonal to the existing `policy` (admission order) axis.

- `preemption_victim="fcfs"` (default): unchanged behavior — FCFS pops the
  last-admitted running request; PRIORITY still evicts
  `max(running, key=(priority, arrival_time))`.
- `preemption_victim="lcf"`: evicts `min(running, key=num_computed_tokens)`,
  the running request with the least sunk compute.

Surface area:

- `vllm/v1/core/sched/request_queue.py`: new `PreemptionVictim` enum
  (`FCFS`, `LCF`), separate from `SchedulingPolicy`.
- `vllm/config/scheduler.py`: `PreemptionVictimPolicy` literal and a
  `preemption_victim: PreemptionVictimPolicy = "fcfs"` field.
- `vllm/v1/core/sched/scheduler.py`: resolve the enum in `__init__` and
  select the victim at the preemption site.
- `vllm/engine/arg_utils.py`: `--preemption-victim` CLI flag, mirroring
  `--scheduling-policy`, plumbed into `SchedulerConfig`.

## Why

Recompute-based preemption resets the victim's `num_computed_tokens` to
zero, so all of its already-computed tokens must be recomputed on resume.
The victim choice therefore controls how much compute a preemption wastes,
and neither existing policy accounts for sunk compute. Evicting the
least-computed running request minimizes that waste. A companion
measurement study observed a **20-59% reduction in wasted recompute tokens**
from least-computed-first victim selection across load and sequence-length
regimes, with no change to admission order.

## Experiments (brief)

We measured the "recompute tax" (fraction of token-compute spent redoing
preempted work) with vLLM on Qwen2.5 models, single A100 unless noted:

- Offline sweep (0.5B), varying the KV-block budget: the tax rises from 0%
  when memory is ample to ~79% under heavy pressure.
- Online serving (7B) via `vllm bench serve` on a real ShareGPT trace with an
  arrival-rate sweep: the tax is 0 at low load and 10-23% under pressure, with
  a clear goodput cliff at overload.
- Scale (14B, 32B via tensor parallelism): the tax persists (15-22%) and LCF's
  benefit holds.
- Ablations (prefix caching, chunked prefill, block size): the tax is largely
  invariant, so it is preemption-driven rather than a config artifact.
- Victim-policy design space (least-computed vs shipped FCFS vs random vs
  most-computed): victim choice alone governs about half the tax.

Across heterogeneous and real-trace loads, LCF cut wasted recompute 20-59% at
no throughput or latency cost; the mechanism is more but cheaper preemptions,
since it evicts near-zero-sunk-cost victims. All runs used public models and
public or self-generated workloads.

## Correctness note (shared bookkeeping)

The LCF victim, like the PRIORITY victim, may be a request that was already
scheduled earlier in the same step, so it cannot use the FCFS `pop()` fast
path. LCF and PRIORITY now share one remove-and-restore path that:

- removes the victim by index (`del self.running[victim_index]`),
- decrements the loop cursor when the victim preceded the current request,
- and, if the victim was already scheduled this step, restores the token
  budget, pops `num_scheduled_tokens` / `req_to_new_blocks` /
  `scheduled_spec_decode_tokens`, and restores the encoder compute budget.

The two policies differ only in the selection key. The FCFS `pop()` path is
untouched, so the default is byte-for-byte the previous behavior.

## Tests

Added to `tests/v1/core/test_scheduler.py`:

- `test_lcf_preemption_victim_selection`: three running requests with
  distinct `num_computed_tokens` under zero-free-block KV pressure; asserts
  LCF preempts the least-computed request (which is neither the request
  being scheduled nor the FCFS victim, so it exercises the shared
  remove+bookkeeping path).
- `test_fcfs_preemption_victim_unchanged`: same setup with the default
  policy; asserts the last-admitted request is still the victim.

Commands and results (CPU, no GPU):

```
VLLM_TARGET_DEVICE=cpu python -m pytest tests/v1/core/test_scheduler.py \
    -k "preempt or priority or lcf" -q
# 27 passed, 8 failed
```

The two new tests and every text-model preemption/priority test pass. The 8
failures are pre-existing multimodal (LlavaForConditionalGeneration) EC
connector / encoder-cache tests that fail at `ModelConfig` model-class
inspection in the constrained local environment, before any scheduler logic
runs; they are unrelated to this change. `ruff check` and `ruff format
--check` pass on all changed files.

## Model evaluation

Not applicable: this change does not alter model outputs, logits, or
sampling. It only changes which already-running request is evicted under KV
pressure; the default keeps existing behavior unchanged.

## Duplicate check

Searched open PRs for the change, not just the keyword. #45558 / #45561 add
a **priority** victim key (a different selection key, orthogonal to this new
axis); #51574 / #41952 are priority-queue re-admission fixes; #49675 handles
zero-progress preemption cascades. No open PR adds a least-computed-first
victim or a `preemption_victim` option (searched "least computed", "lcf",
"preemption_victim", "victim selection"). This is complementary to RFC
#51608 (extensible scheduler-plugin framework): a small default-off built-in
that can ship independently.

## AI assistance disclosure

Per `AGENTS.md`: this change was developed with AI assistance. The human
submitter reviewed every changed line, understands and can defend the change
end-to-end, and ran the tests above.

### CODE REVIEW

**Comment by shivasathishs-rp** [comment]
For the encoder allocation issue I went with avoiding the victim rather than rolling back the allocation. LCF now skips any running request that has an encoder input scheduled in the same step, so we never cancel the worker-side encoder work while leaving the cache reservation behind for a resumed request to falsely hit. The FCFS tail fallback can never hold one either, so that path stays safe. Covered by test_lcf_victim_skips_pending_encoder_allocation.

For the tie case, LCF now excludes the request being scheduled and walks the running queue in reverse, so equal-computed-token requests break the tie toward the last-admitted victim instead of the current request preempting itself into a zero-token step. Covered by test_lcf_victim_avoids_self_preemption_on_tie.

Both new tests pass on CPU. Happy to switch to the explicit allocation rollback instead if you'd rather keep such requests eligible as victims.
